x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java

```diff
@@ -290,6 +334,20 @@ static Step getCurrentStep(PolicyStepsRegistry stepRegistry, String policy, Inde
     static ClusterState moveClusterStateToStep(String indexName, ClusterState currentState, StepKey currentStepKey,
                                                StepKey nextStepKey, LongSupplier nowSupplier,
                                                PolicyStepsRegistry stepRegistry, boolean forcePhaseDefinitionRefresh) {
+        validateTransition(indexName, currentState, currentStepKey, nextStepKey, stepRegistry);
+
+        IndexMetaData idxMeta = currentState.getMetaData().index(indexName);
+        Settings indexSettings = idxMeta.getSettings();
+        String policy = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexSettings);
+        logger.info("moving index [{}] from [{}] to [{}] in policy [{}]",
+            indexName, currentStepKey, nextStepKey, policy);
+
+        return IndexLifecycleRunner.moveClusterStateToNextStep(idxMeta.getIndex(), currentState, currentStepKey,
+            nextStepKey, nowSupplier, forcePhaseDefinitionRefresh);
+    }
+
+    private static void validateTransition(String indexName, ClusterState currentState, StepKey currentStepKey, StepKey nextStepKey,
```

Comment:
Do we have tests for this? Would you mind making this package-visible and adding a unit test for its validation? (Can be done in a separate PR)